pkg/registry/certificates/clustertrustbundle/strategy.go

```diff
@@ -0,0 +1,147 @@
+// Package clustertrustbundle provides Registry interface and its RESTStorage
+// implementation for storing ClusterTrustBundle objects.
+package clustertrustbundle // import "k8s.io/kubernetes/pkg/registry/certificates/clustertrustbundle"
+
+import (
+	"context"
+	"encoding/pem"
+	"fmt"
+	"sort"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/certificates"
+	certvalidation "k8s.io/kubernetes/pkg/apis/certificates/validation"
+	apivalidation "k8s.io/kubernetes/pkg/apis/core/validation"
+)
+
+// strategy implements behavior for ClusterTrustBundles.
+type strategy struct {
+	runtime.ObjectTyper
+	names.NameGenerator
+}
+
+// Strategy is the create, update, and delete strategy for ClusterTrustBundles.
+var Strategy = strategy{legacyscheme.Scheme, names.SimpleNameGenerator}
+
+var _ rest.RESTCreateStrategy = Strategy
+var _ rest.RESTUpdateStrategy = Strategy
+var _ rest.RESTDeleteStrategy = Strategy
+
+func (strategy) NamespaceScoped() bool {
+	return false
+}
+
+func (strategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {}
+
+func noRestrictionsOnName(name string, prefix bool) []string {
+	return nil
+}
+
+func (strategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	bundle := obj.(*certificates.ClusterTrustBundle)
+
+	var allErrors field.ErrorList
+	allErrors = append(allErrors, apivalidation.ValidateObjectMeta(&bundle.ObjectMeta, false, noRestrictionsOnName, field.NewPath("metadata"))...)
+
+	if bundle.Spec.SignerName != "" {
+		signerNameErrors := certvalidation.ValidateSignerName(field.NewPath("spec", "signerName"), bundle.Spec.SignerName)
+		allErrors = append(allErrors, signerNameErrors...)
+	}
+
+	// TODO(KEP-3257): Is it OK to modify the object during validate?
```

Comment:
What do you think we should do here, then?

* PrepareForCreate can't report an error, and the pemTrustAnchors normalization step can encounter lots of errors.
* If we deep-copy the object passed to Validate() so that we don't modify the input object, then we need to re-run the sorting and de-re-serializing of certificates twice (once in PrepareForCreate or Canonicalize, and once in Validate).
* Canonicalize() also can't report an error, so we would need to trust that there's no modification of the object between Validate() and Canonicalize() that would cause pemTrustAnchors normalization to fail.

If we're going to trust what the code calling our strategy does, we might as well trust what it does today, which has no problem with a mutation during Validate() (though we would have to update all the comments that say that Validate() should not mutate the object): https://github.com/kubernetes/kubernetes/blob/66918763265b91000482f82a7e92c68d24733542/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L120

In general, validation and canonicalization can't be cleanly separated without duplicating work.  The current code structure seems a little bit optimistic that they can.

Reply:
I decided to just do the duplication for now.  We can always optimize later.

Reply:
https://github.com/kubernetes/kubernetes/pull/113218#discussion_r1014537571

Reply:
OK, now we just do validation on the API server. The Kubelet half will do any normalization.

Reply:
I think this conversation is resolved.  Validation no longer changes the user's input, just rejects it.